web3indexer/log_processor.py

```diff
@@ -0,0 +1,476 @@
+from datetime import datetime
+
+import structlog
+from web3 import Web3
+
+from .constants import (
+    ERC721_TRANSFER_TOPIC,
+    ERC1155_TRANSFER_SINGLE_TOPIC,
+    ERC1155_TRANSFER_BATCH_TOPIC,
+    ERC_721_IDENTIFIER,
+    ERC_721_METADATA_IDENTIFIER,
+    ERC_1155_IDENTIFIER,
+    ERC_1155_METADATA_IDENTIFIER,
+)
+from .crud import (
+    get_contract,
+    get_nft,
+    upsert_contract,
+    upsert_nft,
+    upsert_transfer,
+)
+from .dispatcher import Dispatcher
+from .models import Contract, ContractType, Nft, Transfer
+from .task import ProcessLogTask
+from .utils import get_nft_id, read_file
+
+
+logger = structlog.get_logger()
+
+ERC165_ABI = read_file("abi/ERC165.json")
+ERC1155_ABI = read_file("abi/ERC1155.json")
+ERC721_ABI = read_file("abi/ERC721.json")
+
+
+class LogProcessor:
+    """
+    Handles fetching, parsing and storing information for a given block.
+    """
+
+    MAX_RETRIES = 5
+
+    def __init__(self, db):
+        self.db = db
+
+    def process_with_retry(
+        self, dispatcher: Dispatcher, w3: Web3, task: ProcessLogTask
+    ):
+        try:
+            self.process(dispatcher, w3, task)
+        except Exception as e:
+            logger.error(e)
+            if task.retries < self.MAX_RETRIES:
+                dispatcher.put(
+                    ProcessLogTask(
```

Comment:
this could cause us to process events in the wrong order, and so we might need to be aware of that when processing the data. for instance, suppose an NFT is transferred from user A to B, then from user B to user C in the same block. if we hit an error processing transfer event 1, we would enqueue this event in the dispatcher, and process event 2 before processing event 1 again. the same issue would happen if there are errors processing the blocks.

the out-of-order processing seems fine for the task of inserting transfer events since we also record the log index. but would we update the NFT data incorrectly if the events are processed in the wrong order?

Reply:
yes! thats a great observation. processing events out of order is fine for tables like `Transfer` but it does matter for how we keep track of ownership. this is something I'm curious about how Rarible handles